daemon: make the snapshotter configurable

thaJeztah opened the pull request:
- upstreaming https://github.com/rumpl/moby/pull/40
- related (will conflict with); https://github.com/moby/moby/pull/43982


Treat (storage/graph)Driver as snapshotter

Also moved some layerStore related initialization to the non-c8d case
because otherwise they get treated as a graphdriver plugins.


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

rumpl committed:
daemon: make the snapshotter configurable

Treat (storage/graph)Driver as snapshotter

Also moved some layerStore related initialization to the non-c8d case
because otherwise they get treated as a graphdriver plugins.

Co-authored-by: Sebastiaan van Stijn <github@gone.nl>
Signed-off-by: Djordje Lukic <djordje.lukic@docker.com>
Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

vvoland commented on daemon/containerd/snapshotters_test.go:
```diff
@@ -0,0 +1,58 @@
+	"gotest.tools/v3/assert"
+)
+
+func TestSnapshotterFromGraphDriver(t *testing.T) {
```
I wonder if this test has any positive value anymore - it made more sense when there was the logic with stripping the fully qualified name (io.containerd.snapshotter.v1).
Now when the function under test is really just one simple switch-case the test is basically a more verbose duplicate of the function that checks if Go's switch-case works as expected 😄

thaJeztah replied:
Yes, it's quite a basic test now; I guess it could still be somewhat useful to test the behaviour for defaults and "other" storage drivers (just in case someone makes changes)

AkihiroSuda commented on daemon/daemon_unix.go:
```diff
@@ -820,7 +820,7 @@ func configureKernelSecuritySupport(config *config.Config, driverName string) er
 		}
 
-		if driverName == "overlay" || driverName == "overlay2" {
+		if driverName == "overlay" || driverName == "overlay2" || driverName == "overlayfs" {
```
This is likely to cause confusions to the end users.

I think the containerd snapshotter name should be mapped to something like `containerd://overlayfs` or `containerd:overlayfs`

thaJeztah replied:
There was some discussion about this in a maintainers call; see https://github.com/rumpl/moby/pull/40#discussion_r938221864

> We discussed this in the maintainers meeting;
> 
> * concensus was that the preference is to re-use the `--storage-driver` flag, as the term is generic enough, and was originally chosen for that reason (apparently there had been some internal discussion at that time to get rid of the "layers" concept, and early implementations of "snapshotter-like approaches" were already being designed)
> * for _now_, we would switch between "containerd snapshotters" and "storage drivers" based on the `feature` flag that was added (but snapshotter would of course at some point become the default)
> * if that feature is enabled, we must have some mapping for backward compatibility on (already familiar) names that have an equivalent in containerd, so;
>   
>   * `overlay` -> `io.containerd.snapshotter.v1.overlayfs`
>   * `overlay2` -> `io.containerd.snapshotter.v1.overlayfs` (same)
>   * `zfs` -> `io.containerd.snapshotter.v1.zfs`
>   * etc.
> * besides the "shorthand" (`overlay2`) names, also accept the "fully qualified" names (e.g. `io.containerd.snapshotter.v1.overlayfs` can be used in the daemon config)
> * if no explicit snapshotter is defined in the daemon config we can use containerd's default
> * but, we should consider having a better automatic selection (similar to how we currently select the most suitable snapshotter) instead of containerd's (currently hardcoded) default; this could be by querying (if possible) the list of available snapshotter-plugins that were loaded, and pick the first available one (based on a priority list).
> * perhaps this could use a contribution to containerd, to make containerd itself also make a similar selection (TBD)

thaJeztah replied:
Overall it surprised me a bit that containerd didn't accept the full `io.containerd.snapshotter.v1.overlayfs` (which could've been an alternative). 

That said; the point of view there was that both "old" and "new" (containerd-snapshotter) being supported was more of a transitional period, after which there would be only "one" option.

Is your concern the configuration side of things, or what's shown in `docker info`? (perhaps for the `docker info` case we could consider having a different presentation)

/cc @tianon @cpuguy83 @tonistiigi @corhere any thoughts?

thaJeztah replied:
^^ I should mention that we dropped the `io.containerd.snapshotter.v1.` handling in that PR, as it wouldn't (currently) match what's accepted by containerd, but perhaps that's a part that could be re-considered (and perhaps added in containerd itself)

AkihiroSuda replied:
> (perhaps for the `docker info` case we could consider having a different presentation)

SGTM

AkihiroSuda replied:
The documentation will also have to emphasize that `overlayfs` is actually newer than `overlay2`

thaJeztah replied:
Yes; documentation will be added specifically about the containerd snapshotter integration; once all bits are integrated here in upstream, we will also update the "storage driver" section to describe them.

thaJeztah commented:
rebased, because https://github.com/moby/moby/pull/43982 was merged

thaJeztah committed:
info: add driver-type

With this patch:

    mkdir -p /etc/docker/
    echo '{"features":{"containerd-snapshotter":true}}' > /etc/docker/daemon.json
    dockerd

    docker info
    ...
    Storage Driver: overlayfs
     driver-type: io.containerd.snapshotter.v1
    Logging Driver: json-file

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on daemon/containerd/service.go:
```diff
@@ -62,7 +64,10 @@ func (i *ImageService) GetLayerByID(cid string) (layer.RWLayer, error) {
-	return [][2]string{}
+	// TODO(thaJeztah) do we want to add more details about the driver here?
+	return [][2]string{
+		{"driver-type", "io.containerd.snapshotter.v1"},
```
Or do we want to use the `plugins.SnapshotPlugin` const for this? https://github.com/containerd/containerd/blob/902212651b12e4d81b7bc779632b2e4da0a8e673/plugin/plugin.go#L65-L66

```go
	// SnapshotPlugin implements a snapshotter
	SnapshotPlugin Type = "io.containerd.snapshotter.v1"
```

thaJeztah replied:
Ah, whoops, forgot to cast it to a string;

```
# github.com/docker/docker/daemon/containerd
..\..\daemon\containerd\service.go:70:19: cannot use plugin.SnapshotPlugin (constant "io.containerd.snapshotter.v1" of type plugin.Type) as type string in array or slice literal
```

thaJeztah commented:
@tianon @tonistiigi @corhere @cpuguy83 this one LGTY?

thaJeztah commented:
@AkihiroSuda this LGTY?

thaJeztah commented:
Thanks! For others; I think this matches what was discussed, so I'll bring this in. If there's changes to be made, I'm happy to do so in a follow up (this PR will only be in master / v-next, so we'll have time to tweak if needed).
